### PR TITLE
refactor(crypto): improve package structure

### PR DESCRIPTION
**Pull request by ctrlc03**
# Description

improve the organization of the package by further separating classes and functions

## Confirmation

- [x] I have read and understand MACI's [contributor guidelines](../CONTRIBUTING.md) and [code of conduct requirements](../CODE_OF_CONDUCT.md).
- [x] I have read and understand MACI's [GitHub processes](https://github.com/privacy-scaling-explorations/maci/discussions/847) documentation.
- [x] I have read and understand MACI's [testing guide](https://maci.pse.dev/docs/testing).

### CODE REVIEW

File: crypto/ts/hashing.ts [hunk 1]
```diff
@@ -0,0 +1,146 @@
+import { poseidon } from "circomlib";
+import * as ethers from "ethers";
```
Comment (0xmad) [inline crypto/ts/hashing.ts:2]:
```suggestion
import { utils } from "ethers";
```

File: crypto/ts/hashing.ts [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+  const elementLength = elements.length;
+  if (elements.length > numElements) {
+    throw new TypeError(`the length of the elements array should be at most ${numElements}; got ${elements.length}`);
+  }
+  const elementsPadded = elements.slice();
+  if (elementLength < numElements) {
+    for (let i = elementLength; i < numElements; i += 1) {
+      elementsPadded.push(BigInt(0));
+    }
+  }
```
Comment (0xmad) [inline crypto/ts/hashing.ts:96]:
```suggestion
  const elementLength = elements.length;

  if (elements.length > numElements) {
    throw new TypeError(`the length of the elements array should be at most ${numElements}; got ${elements.length}`);
  }

  const elementsPadded = elements.slice();

  if (elementLength < numElements) {
    for (let i = elementLength; i < numElements; i += 1) {
      elementsPadded.push(BigInt(0));
    }
  }
```

File: crypto/ts/hashing.ts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+  return poseidonT6([
+    elementsPadded[0],
+    poseidonT6(elementsPadded.slice(1, 6)),
+    poseidonT6(elementsPadded.slice(6, 11)),
+    elementsPadded[11],
+    elementsPadded[12],
+  ]);
+};
```
Comment (0xmad) [inline crypto/ts/hashing.ts:139]:
```suggestion
export const hash13 = (elements: Plaintext): bigint => {
  const max = 13;
  const elementLength = elements.length;

  if (elementLength > max) {
    throw new TypeError(`the length of the elements array should be at most ${max}; got ${elements.length}`);
  }

  const elementsPadded = elements.slice();

  if (elementLength < max) {
    for (let i = elementLength; i < max; i += 1) {
      elementsPadded.push(BigInt(0));
    }
  }

  return poseidonT6([
    elementsPadded[0],
    poseidonT6(elementsPadded.slice(1, 6)),
    poseidonT6(elementsPadded.slice(6, 11)),
    elementsPadded[11],
    elementsPadded[12],
  ]);
};
```

**Review by 0xmad — COMMENTED** [review]
@ctrlc03 thanks, just few comments

**Review by 0xmad — APPROVED** [review]
@ctrlc03 thanks!

File: crypto/ts/babyjub.ts [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+   * @returns the point as a contract param
+   */
+  asContractParam(): { x: string[]; y: string[] } {
+    return {
+      x: this.x.map((n) => n.toString()),
+      y: this.y.map((n) => n.toString()),
+    };
+  }
+
+  private checkPointsRange(x: bigint[], type: "x" | "y") {
```
[the remark this answers is not in the document]
Comment (ctrlc03) [reply crypto/ts/babyjub.ts:97]: done

**Pull request merged**